File: src/app/core/services/billing.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of, throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import {
  BillingSubscription,
  Invoice,
  PaymentMethod,
  UsageMetrics,
} from '../models/organization.model';
import { LoggingService } from './logging.service';

interface ApiResponse<T> {
  success: boolean;
  data: T;
  error?: string;
}

@Injectable({
  providedIn: 'root',
})
export class BillingService {
  private http = inject(HttpClient);
  private logging = inject(LoggingService);

  getSubscription(organizationId: string): Observable<BillingSubscription> {
    return this.http
      .get<ApiResponse<any>>(`/api/org/${organizationId}/dashboard`)
      .pipe(
        map((response) => {
          const now = new Date();
          return {
            id: `sub_${organizationId}`,
            organizationId,
            stripeSubscriptionId: undefined,
            planId: 'plan_pro',
            planName: 'Pro',
            status: 'active',
            currentPeriodStart: now,
            currentPeriodEnd: new Date(
              now.getTime() + 30 * 24 * 60 * 60 * 1000,
            ),
            autoRenew: true,
            createdAt: now,
            updatedAt: now,
          } as BillingSubscription;
        }),
        catchError((error) =>
          throwError(
            () => new Error(error?.error?.detail || 'Subscription not found'),
          ),
        ),
      );
  }

  createSubscription(
    organizationId: string,
    planId: string,
  ): Observable<BillingSubscription> {
    return this.http
      .post<
        ApiResponse<any>
      >(`/api/billing/subscriptions`, { organizationId, planId })
      .pipe(
        map((response) =>
          this.mapSubscription(response.data, organizationId, planId),
        ),
        catchError((error) =>
          throwError(
            () =>
              new Error(
                error?.error?.detail || 'Failed to create subscription',
              ),
          ),
        ),
      );
  }

  changePlan(
    subscriptionId: string,
    newPlanId: string,
    prorationBehavior: string = 'create_prorations',
  ): Observable<BillingSubscription> {
    return this.http
      .patch<ApiResponse<any>>(`/api/billing/subscriptions/${subscriptionId}`, {
        planId: newPlanId,
        prorationBehavior,
      })
      .pipe(
        map((response) =>
          this.mapSubscription(
            response.data,
            response.data?.organizationId,
            newPlanId,
            subscriptionId,
          ),
        ),
        catchError((error) =>
          throwError(
            () => new Error(error?.error?.detail || 'Failed to change plan'),
          ),
        ),
      );
  }

  cancelSubscription(
    subscriptionId: string,
    reason?: string,
    immediate: boolean = false,
  ): Observable<BillingSubscription> {
    return this.http
      .post<ApiResponse<any>>(
        `/api/billing/subscriptions/${subscriptionId}/cancel`,
        {
          reason,
          immediate,
        },
      )
      .pipe(
        map((response) =>
          this.mapSubscription(
            response.data,
            response.data?.organizationId,
            response.data?.planId,
            subscriptionId,
          ),
        ),
        catchError((error) =>
          throwError(
            () =>
              new Error(
                error?.error?.detail || 'Failed to cancel subscription',
              ),
          ),
        ),
      );
  }

  pauseSubscription(
    subscriptionId: string,
    resumeAt?: Date,
  ): Observable<BillingSubscription> {
    return this.http
      .post<ApiResponse<any>>(
        `/api/billing/subscriptions/${subscriptionId}/pause`,
        {
          resumeAt: resumeAt?.toISOString(),
        },
      )
      .pipe(
        map((response) =>
          this.mapSubscription(
            response.data,
            response.data?.organizationId,
            response.data?.planId,
            subscriptionId,
          ),
        ),
        catchError((error) =>
          throwError(
            () =>
              new Error(error?.error?.detail || 'Failed to pause subscription'),
          ),
        ),
      );
  }

  resumeSubscription(subscriptionId: string): Observable<BillingSubscription> {
    return this.http
      .post<
        ApiResponse<any>
      >(`/api/billing/subscriptions/${subscriptionId}/resume`, {})
      .pipe(
        map((response) =>
          this.mapSubscription(
            response.data,
            response.data?.organizationId,
            response.data?.planId,
            subscriptionId,
          ),
        ),
        catchError((error) =>
          throwError(
            () =>
              new Error(
                error?.error?.detail || 'Failed to resume subscription',
              ),
          ),
        ),
      );
  }

  getInvoices(
    organizationId: string,
    options?: { limit?: number; offset?: number },
  ): Observable<{ data: Invoice[]; total: number; hasMore: boolean }> {
    const limit = options?.limit || 10;
    const offset = options?.offset || 0;

    return this.http
      .get<
        ApiResponse<{ invoices?: any[]; total?: number }>
      >(`/api/billing/invoices?organizationId=${organizationId}&limit=${limit}&offset=${offset}`)
      .pipe(
        map((response) => {
          const invoices = (response.data?.invoices || []).map((invoice) =>
            this.mapInvoice(invoice, organizationId),
          );
          const total = response.data?.total || invoices.length;
          return {
            data: invoices,
            total,
            hasMore: offset + limit < total,
          };
        }),
        catchError(() => of({ data: [], total: 0, hasMore: false })),
      );
  }

  getInvoice(id: string): Observable<Invoice> {
    return this.http.get<ApiResponse<any>>(`/api/billing/invoices/${id}`).pipe(
      map((response) =>
        this.mapInvoice(
          response.data,
          response.data?.organizationId || 'unknown',
        ),
      ),
      catchError((error) =>
        throwError(
          () => new Error(error?.error?.detail || 'Invoice not found'),
        ),
      ),
    );
  }

  downloadInvoicePDF(
    id: string,
  ): Observable<{ url: string; fileName: string }> {
    return this.http
      .get<
        ApiResponse<{ url?: string; fileName?: string }>
      >(`/api/billing/invoices/${id}/pdf`)
      .pipe(
        map((response) => ({
          url: response.data?.url || `/api/invoices/${id}/pdf`,
          fileName: response.data?.fileName || `invoice_${id}.pdf`,
        })),
        catchError((error) =>
          throwError(
            () =>
              new Error(
                error?.error?.detail || 'Failed to download invoice PDF',
              ),
          ),
        ),
      );
  }

  getPaymentMethods(organizationId: string): Observable<PaymentMethod[]> {
    return this.http
      .get<
        ApiResponse<{ paymentMethods?: any[] }>
      >(`/api/billing/payment-methods?organizationId=${organizationId}`)
      .pipe(
        map((response) =>
          (response.data?.paymentMethods || []).map((method) =>
            this.mapPaymentMethod(method, organizationId),
          ),
        ),
        catchError(() => of([])),
      );
  }

  addPaymentMethod(
    organizationId: string,
    stripeToken: string,
    setAsDefault: boolean = false,
  ): Observable<PaymentMethod> {
    return this.http
      .post<ApiResponse<any>>(`/api/billing/payment-methods`, {
        organizationId,
        stripeToken,
        setAsDefault,
      })
      .pipe(
        map((response) => this.mapPaymentMethod(response.data, organizationId)),
        catchError((error) =>
          throwError(
            () =>
              new Error(error?.error?.detail || 'Failed to add payment method'),
          ),
        ),
      );
  }

  deletePaymentMethod(id: string): Observable<{ success: boolean }> {
    return this.http
      .delete<ApiResponse<any>>(`/api/billing/payment-methods/${id}`)
      .pipe(
        map(() => ({ success: true })),
        catchError((error) =>
          throwError(
            () =>
              new Error(
                error?.error?.detail || 'Failed to delete payment method',
              ),
          ),
        ),
      );
  }

  getUsageMetrics(organizationId: string): Observable<UsageMetrics> {
    return this.http
      .get<ApiResponse<any>>(`/api/org/${organizationId}/dashboard`)
      .pipe(
        map((response) => {
          const kpis = response.data?.dashboard?.kpis || {};
          return {
            id: `usage_${organizationId}`,
            organizationId,
            month: new Date().toISOString().substring(0, 7),
            apiCallsUsed: Number(kpis.api_calls_this_month || 0),
            apiCallsLimit: 1000000,
            usersUsed: Number(kpis.total_users || 0),
            usersLimit: 999,
            teamsUsed: Number(kpis.active_features || 0),
            teamsLimit: 999,
            storageUsed: 0,
            storageLimit: 100,
            webhooksUsed: 0,
            webhooksLimit: 100,
            lastUpdatedAt: new Date(),
          };
        }),
        catchError((error) =>
          throwError(
            () =>
              new Error(
                error?.error?.detail || 'Failed to fetch usage metrics',
              ),
          ),
        ),
      );
  }

  getBillingStats(organizationId: string): Observable<{
    totalInvoices: number;
    paidInvoices: number;
    outstandingInvoices: number;
    totalRevenue: number;
    outstandingAmount: number;
    averageInvoiceAmount: number;
  }> {
    return this.getInvoices(organizationId, { limit: 100, offset: 0 }).pipe(
      map(({ data }) => {
        const paid = data.filter((inv) => inv.status === 'paid');
        const outstanding = data.filter((inv) => inv.status !== 'paid');
        const totalRevenue = paid.reduce((sum, inv) => sum + inv.amount, 0);
        const outstandingAmount = outstanding.reduce(
          (sum, inv) => sum + inv.amount,
          0,
        );

        return {
          totalInvoices: data.length,
          paidInvoices: paid.length,
          outstandingInvoices: outstanding.length,
          totalRevenue,
          outstandingAmount,
          averageInvoiceAmount:
            paid.length > 0 ? totalRevenue / paid.length : 0,
        };
      }),
    );
  }

  private mapSubscription(
    data: any,
    organizationId: string,
    planId: string,
    subscriptionId?: string,
  ): BillingSubscription {
    const now = new Date();
    const currentPlanId = planId || data?.planId || 'plan_pro';

    return {
      id: data?.id || subscriptionId || `sub_${Date.now()}`,
      organizationId: organizationId || data?.organizationId || 'unknown',
      stripeSubscriptionId: data?.stripeSubscriptionId,
      planId: currentPlanId,
      planName: data?.planName || this.getPlanName(currentPlanId),
      status: data?.status || 'active',
      currentPeriodStart: data?.currentPeriodStart
        ? new Date(data.currentPeriodStart)
        : now,
      currentPeriodEnd: data?.currentPeriodEnd
        ? new Date(data.currentPeriodEnd)
        : new Date(now.getTime() + 30 * 24 * 60 * 60 * 1000),
      trialEndsAt: data?.trialEndsAt ? new Date(data.trialEndsAt) : undefined,
      canceledAt: data?.canceledAt ? new Date(data.canceledAt) : undefined,
      cancelledReason: data?.cancelledReason,
      autoRenew: data?.autoRenew ?? true,
      createdAt: data?.createdAt ? new Date(data.createdAt) : now,
      updatedAt: data?.updatedAt ? new Date(data.updatedAt) : now,
    };
  }

  private mapInvoice(data: any, organizationId: string): Invoice {
    return {
      id: data?.id || `inv_${Date.now()}`,
      organizationId,
      stripeInvoiceId: data?.stripeInvoiceId,
      subscriptionId: data?.subscriptionId || '',
      amount: Number(data?.amount || 0),
      currency: data?.currency || 'USD',
      status: data?.status || 'draft',
      issuedAt: data?.issuedAt ? new Date(data.issuedAt) : new Date(),
      dueAt: data?.dueAt ? new Date(data.dueAt) : new Date(),
      paidAt: data?.paidAt ? new Date(data.paidAt) : undefined,
      lineItems: data?.lineItems || [],
      metadata: data?.metadata,
      createdAt: data?.createdAt ? new Date(data.createdAt) : new Date(),
      updatedAt: data?.updatedAt ? new Date(data.updatedAt) : new Date(),
    };
  }

  private mapPaymentMethod(data: any, organizationId: string): PaymentMethod {
    return {
      id: data?.id || `pm_${Date.now()}`,
      organizationId,
      stripePaymentMethodId: data?.stripePaymentMethodId,
      type: data?.type || 'card',
      cardBrand: data?.cardBrand,
      cardLast4: data?.cardLast4,
      cardExpMonth: data?.cardExpMonth,
      cardExpYear: data?.cardExpYear,
      status: data?.status || 'active',
      isDefault: data?.isDefault ?? false,
      createdAt: data?.createdAt ? new Date(data.createdAt) : new Date(),
      updatedAt: data?.updatedAt ? new Date(data.updatedAt) : new Date(),
    };
  }

  private getPlanName(planId: string): string {
    const names: Record<string, string> = {
      plan_starter: 'Starter',
      plan_pro: 'Pro',
      plan_enterprise: 'Enterprise',
    };
    return names[planId] || 'Unknown';
  }
}

File: backend/routes/billing.py
"""
Billing Routes
Handles subscriptions, invoices, payment methods, and billing operations via Stripe
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from datetime import datetime
from typing import Optional, List
import logging

from ..models import ApiResponse, User
from ..dependencies import get_current_user
from ..stripe_service import stripe_service
from ..database import get_firestore_db

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/billing", tags=["billing"])


@router.get("/plans", response_model=ApiResponse)
async def get_billing_plans():
    """Get list of available billing plans/prices"""
    try:
        prices = stripe_service.list_prices()
        
        plans = []
        for price in prices:
            plans.append({
                "id": price['id'],
                "name": price['product_name'],
                "amount": price['amount'],
                "currency": price['currency'],
                "billing_period": price['billing_period'],
                "interval_count": price['interval_count'],
                "created_at": datetime.fromtimestamp(price['created_at']),
            })
        
        return ApiResponse(
            success=True,
            data={"plans": sorted(plans, key=lambda x: x['amount'])}
        )
    except Exception as e:
        logger.error(f"Failed to list billing plans: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/subscriptions", response_model=ApiResponse)
async def get_subscriptions(
    organization_id: str = Query(...),
    current_user: User = Depends(get_current_user)
):
    """Get subscriptions for organization"""
    try:
        # TODO: Verify user has access to this organization
        firestore_db = get_firestore_db()
        
        org_doc = firestore_db.collection('organizations').document(organization_id).get()
        if not org_doc.exists:
            raise HTTPException(status_code=404, detail="Organization not found")
        
        org_data = org_doc.to_dict()
        stripe_customer_id = org_data.get('stripe_customer_id')
        
        if not stripe_customer_id:
            return ApiResponse(success=True, data={"subscriptions": []})
        
        # List subscriptions for customer
        stripe_client = stripe_service.get_client()
        subscriptions = stripe_client.Subscription.list(customer=stripe_customer_id)
        
        subs = []
        for sub in subscriptions['data']:
            subs.append({
                "id": sub.id,
                "status": sub.status,
                "price_id": sub['items']['data'][0]['price']['id'] if sub['items']['data'] else None,
                "current_period_start": datetime.fromtimestamp(sub.current_period_start),
                "current_period_end": datetime.fromtimestamp(sub.current_period_end),
                "auto_renew": True,
                "created_at": datetime.fromtimestamp(sub.created),
                "canceled_at": datetime.fromtimestamp(sub.canceled_at) if sub.canceled_at else None,
            })
        
        return ApiResponse(success=True, data={"subscriptions": subs})
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to get subscriptions: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/subscriptions", response_model=ApiResponse)
async def create_subscription(
    organization_id: str = Query(...),
    price_id: str = Query(...),
    current_user: User = Depends(get_current_user)
):
    """Create subscription for organization"""
    try:
        firestore_db = get_firestore_db()
        
        org_doc = firestore_db.collection('organizations').document(organization_id).get()
        if not org_doc.exists:
            raise HTTPException(status_code=404, detail="Organization not found")
        
        org_data = org_doc.to_dict()
        stripe_customer_id = org_data.get('stripe_customer_id')
        
        # Create customer if doesn't exist
        if not stripe_customer_id:
            customer_info = stripe_service.create_customer(
                email=org_data.get('billing_email', f"billing@{organization_id}.local"),
                metadata={
                    'organization_id': organization_id,
                    'organization_name': org_data.get('name', ''),
                }
            )
            stripe_customer_id = customer_info['id']
            
            # Update org with customer ID
            firestore_db.collection('organizations').document(organization_id).update({
                'stripe_customer_id': stripe_customer_id,
                'updated_at': datetime.now(),
            })
        
        # Create subscription
        subscription = stripe_service.create_subscription(
            customer_id=stripe_customer_id,
            price_id=price_id,
            metadata={'organization_id': organization_id}
        )
        
        return ApiResponse(
            success=True,
            data={
                "subscription": {
                    "id": subscription['id'],
                    "status": subscription['status'],
                    "price_id": subscription['plan_id'],
                    "current_period_start": subscription['created_at'],
                    "current_period_end": subscription['current_period_end'],
                    "created_at": subscription['created_at'],
                }
            }
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to create subscription: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.patch("/subscriptions/{subscription_id}", response_model=ApiResponse)
async def update_subscription(
    subscription_id: str,
    price_id: Optional[str] = Query(None),
    status: Optional[str] = Query(None),  # resume, cancel (soft)
    current_user: User = Depends(get_current_user)
):
    """Update subscription (change price, pause, resume)"""
    try:
        subscription = stripe_service.get_subscription(subscription_id)
        
        if status == 'resume':
            # Resume paused subscription
            stripe_client = stripe_service.get_client()
            stripe_subscription = stripe_client.Subscription.retrieve(subscription_id)
            if stripe_subscription.pause_collection:
                updated = stripe_client.Subscription.modify(
                    subscription_id,
                    pause_collection=None
                )
                subscription = stripe_service._map_subscription(updated)
        
        elif price_id:
            # Change price/plan
            subscription = stripe_service.update_subscription(
                subscription_id,
                price_id=price_id
            )
        
        return ApiResponse(
            success=True,
            data={"subscription": {
                "id": subscription['id'],
                "status": subscription['status'],
                "price_id": subscription['plan_id'],
                "current_period_start": subscription['current_period_start'],
                "current_period_end": subscription['current_period_end'],
            }}
        )
    except Exception as e:
        logger.error(f"Failed to update subscription: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/subscriptions/{subscription_id}/cancel", response_model=ApiResponse)
async def cancel_subscription(
    subscription_id: str,
    immediate: bool = Query(False),
    current_user: User = Depends(get_current_user)
):
    """Cancel subscription"""
    try:
        subscription = stripe_service.cancel_subscription(
            subscription_id,
            immediate=immediate
        )
        
        return ApiResponse(
            success=True,
            data={
                "subscription": {
                    "id": subscription['id'],
                    "status": subscription['status'],
                    "canceled_at": subscription['canceled_at'],
                }
            }
        )
    except Exception as e:
        logger.error(f"Failed to cancel subscription: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/invoices", response_model=ApiResponse)
async def get_invoices(
    organization_id: str = Query(...),
    limit: int = Query(10, ge=1, le=100),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user)
):
    """Get invoices for organization"""
    try:
        firestore_db = get_firestore_db()
        
        org_doc = firestore_db.collection('organizations').document(organization_id).get()
        if not org_doc.exists:
            raise HTTPException(status_code=404, detail="Organization not found")
        
        org_data = org_doc.to_dict()
        stripe_customer_id = org_data.get('stripe_customer_id')
        
        if not stripe_customer_id:
            return ApiResponse(success=True, data={"invoices": [], "total": 0})
        
        invoices = stripe_service.list_invoices(stripe_customer_id, limit=limit)
        
        invoice_list = []
        for inv in invoices:
            invoice_list.append({
                "id": inv['id'],
                "amount": inv['amount'],
                "currency": inv['currency'],
                "status": inv['status'],
                "issued_at": inv['issued_at'],
                "paid_at": inv['paid_at'],
                "due_at": inv['due_at'],
                "invoice_pdf": inv['invoice_pdf'],
                "hosted_invoice_url": inv['hosted_invoice_url'],
            })
        
        return ApiResponse(
            success=True,
            data={
                "invoices": invoice_list[offset:offset+limit],
                "total": len(invoice_list),
            }
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to get invoices: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/invoices/{invoice_id}/pdf", response_model=ApiResponse)
async def get_invoice_pdf(
    invoice_id: str,
    current_user: User = Depends(get_current_user)
):
    """Get invoice PDF URL"""
    try:
        invoice = stripe_service.get_invoice(invoice_id)
        
        return ApiResponse(
            success=True,
            data={
                "pdf_url": invoice['invoice_pdf'],
                "hosted_url": invoice['hosted_invoice_url'],
            }
        )
    except Exception as e:
        logger.error(f"Failed to get invoice PDF: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/payment-methods", response_model=ApiResponse)
async def get_payment_methods(
    organization_id: str = Query(...),
    current_user: User = Depends(get_current_user)
):
    """Get payment methods for organization"""
    try:
        firestore_db = get_firestore_db()
        
        org_doc = firestore_db.collection('organizations').document(organization_id).get()
        if not org_doc.exists:
            raise HTTPException(status_code=404, detail="Organization not found")
        
        org_data = org_doc.to_dict()
        stripe_customer_id = org_data.get('stripe_customer_id')
        
        if not stripe_customer_id:
            return ApiResponse(success=True, data={"payment_methods": []})
        
        methods = stripe_service.list_payment_methods(stripe_customer_id)
        
        payment_methods = []
        for method in methods:
            payment_methods.append({
                "id": method['id'],
                "type": method['type'],
                "card_brand": method['card_brand'],
                "card_last4": method['card_last4'],
                "card_exp_month": method['card_exp_month'],
                "card_exp_year": method['card_exp_year'],
                "created_at": method['created_at'],
            })
        
        return ApiResponse(success=True, data={"payment_methods": payment_methods})
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to get payment methods: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/payment-methods", response_model=ApiResponse)
async def add_payment_method(
    organization_id: str = Query(...),
    payment_method_id: str = Query(...),
    set_default: bool = Query(False),
    current_user: User = Depends(get_current_user)
):
    """Add payment method to organization"""
    try:
        firestore_db = get_firestore_db()
        
        org_doc = firestore_db.collection('organizations').document(organization_id).get()
        if not org_doc.exists:
            raise HTTPException(status_code=404, detail="Organization not found")
        
        org_data = org_doc.to_dict()
        stripe_customer_id = org_data.get('stripe_customer_id')
        
        if not stripe_customer_id:
            raise HTTPException(status_code=400, detail="Organization has no Stripe customer")
        
        # Attach payment method to customer
        method = stripe_service.attach_payment_method(
            payment_method_id,
            stripe_customer_id
        )
        
        # Set as default if requested
        if set_default:
            stripe_client = stripe_service.get_client()
            stripe_client.Customer.modify(
                stripe_customer_id,
                invoice_settings={"default_payment_method": payment_method_id}
            )
        
        return ApiResponse(
            success=True,
            data={
                "payment_method": {
                    "id": method['id'],
                    "type": method['type'],
                    "card_brand": method['card_brand'],
                    "card_last4": method['card_last4'],
                    "card_exp_month": method['card_exp_month'],
                    "card_exp_year": method['card_exp_year'],
                }
            }
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to add payment method: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/payment-methods/{payment_method_id}", response_model=ApiResponse)
async def remove_payment_method(
    payment_method_id: str,
    current_user: User = Depends(get_current_user)
):
    """Remove payment method from organization"""
    try:
        stripe_service.detach_payment_method(payment_method_id)
        
        return ApiResponse(
            success=True,
            data={"message": "Payment method removed"}
        )
    except Exception as e:
        logger.error(f"Failed to remove payment method: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))
